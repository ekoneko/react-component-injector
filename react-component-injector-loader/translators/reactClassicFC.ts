/**
 * React translate jsx with React.createElement before 17.
 * It need import react manually.
 */

import * as t from "@babel/types";
import { addInjectorLib, checkInjectorLibExists, wrapVariable } from "../utils";

export function getReactTag(programNode: t.Program) {
  const reactDeclarationNode = programNode.body.find(
    (node) =>
      node.type === "ImportDeclaration" &&
      node.source.value === "react"
  ) as t.ImportDeclaration | undefined;
  const reactDefaultSpecifier = reactDeclarationNode?.specifiers.find(specifier => specifier.type === 'ImportDefaultSpecifier') as t.ImportDefaultSpecifier | null
  return reactDefaultSpecifier?.local.name
}

export function wrapReactClassicFC(programNode: t.Program, { reactTag, resourceRelativePath, encrypt }: { reactTag: string; resourceRelativePath: string; encrypt: (key: string) => string }) {
  let injectorLibExists = checkInjectorLibExists(programNode);
  programNode.body.forEach((node, index) => {
    if (node.type !== "ExportNamedDeclaration") {
      return;
    }
    const variableDeclarationNode = node.declaration;
    if (
      variableDeclarationNode?.type !== "VariableDeclaration" ||
      variableDeclarationNode.declarations[0]?.type !==
      "VariableDeclarator" ||
      variableDeclarationNode.declarations[0]?.init?.type !==
      "FunctionExpression"
    ) {
      return;
    }
    const returnStatementNode =
      variableDeclarationNode.declarations[0]?.init.body.body.find(
        (item) => item.type === "ReturnStatement"
      ) as t.ReturnStatement;
    if (returnStatementNode?.argument?.type !== "CallExpression") {
      return;
    }
    if (returnStatementNode?.argument?.type !== "CallExpression") {
      return;
    }
    const returnCallee = returnStatementNode.argument.callee;
    if (
      returnCallee.type !== 'MemberExpression' ||
      returnCallee.object.type !== 'Identifier' ||
      returnCallee.object.name !== reactTag ||
      returnCallee.property.type !== 'Identifier' ||
      returnCallee.property.name !== 'createElement' ||
      variableDeclarationNode.declarations[0].id.type !== "Identifier"
    ) {
      return;
    }
    const componentName = variableDeclarationNode.declarations[0].id.name;
    if (!injectorLibExists) {
      addInjectorLib(programNode);
      index += 1;
      injectorLibExists = true;
    }
    programNode.body.splice(index, 1, variableDeclarationNode, ...wrapVariable(componentName, encrypt(`${resourceRelativePath}#${componentName}`)));
  });
}