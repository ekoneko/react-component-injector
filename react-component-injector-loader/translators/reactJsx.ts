/**
 * React-jsx is a new way to parse jsx code which not need import React.
 * On this scene, jsx file will auto import "react/jsx-runtime" 
 * and transform jsx to "_jsx("div", { children: "Text" })"
 */
import * as t from "@babel/types";
import { addInjectorLib, checkInjectorLibExists, wrapVariable } from '../utils'

export function getJsxRuntimeTag(programNode: t.Program) {
  const jsxRuntimeDeclarationNode = programNode.body.find(
    (node) =>
      node.type === "ImportDeclaration" &&
      node.source.value === "react/jsx-runtime"
  ) as t.ImportDeclaration | undefined;
  return (
    // react/jsx-runtime is added automated, the type never changed.
    (
      jsxRuntimeDeclarationNode?.specifiers as [t.ImportSpecifier] | undefined
    )?.[0].local.name
  );
}

export function wrapReactJsxComponent(programNode: t.Program, { jsxTag, resourceRelativePath, encrypt }: { jsxTag: string; resourceRelativePath: string; encrypt: (key: string) => string }) {
  let injectorLibExists = checkInjectorLibExists(programNode);
  programNode.body.forEach((node, index) => {
    if (node.type !== "ExportNamedDeclaration") {
      return;
    }
    const variableDeclarationNode = node.declaration;
    if (
      variableDeclarationNode?.type !== "VariableDeclaration" ||
      variableDeclarationNode.declarations[0]?.type !==
      "VariableDeclarator" ||
      variableDeclarationNode.declarations[0]?.init?.type !==
      "FunctionExpression"
    ) {
      return;
    }
    const returnStatementNode =
      variableDeclarationNode.declarations[0]?.init.body.body.find(
        (item) => item.type === "ReturnStatement"
      ) as t.ReturnStatement;
    if (returnStatementNode?.argument?.type !== "CallExpression") {
      return;
    }
    const returnCallee = returnStatementNode.argument.callee;
    if (
      returnCallee.type !== "Identifier" ||
      returnCallee.name !== jsxTag ||
      variableDeclarationNode.declarations[0].id.type !== "Identifier"
    ) {
      return;
    }
    const componentName = variableDeclarationNode.declarations[0].id.name;
    if (!injectorLibExists) {
      addInjectorLib(programNode);
      index += 1;
      injectorLibExists = true;
    }
    programNode.body.splice(index, 1, variableDeclarationNode, ...wrapVariable(componentName, encrypt(`${resourceRelativePath}#${componentName}`)));
  });
}