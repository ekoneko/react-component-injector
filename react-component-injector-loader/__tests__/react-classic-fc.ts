import { parse as babelParse } from "@babel/parser";
import reactInjectorLoader from '../index'
import { getReactTag } from "../translators/reactClassicFC";

const code = `import React from 'react';
export var Breadcrumb = function (_a) {
  return (React.createElement(StyleWrapper, { className: className },
    React.createElement(StyleContainer, null, items.map(function (item, index) {
        if (index !== 0 && index < hiddenCount + 1) {
            return null;
        }
        return (React.createElement(StyleItem, { key: index },
            React.createElement(StyleText, { onClick: createHandleClick(index), width: index < collapsedCount ? collapsedWidthPx + "px" : undefined }, item),
            React.createElement(StyleArrowWrapper, null, arrowIcon !== null && arrowIcon !== void 0 ? arrowIcon : React.createElement(StyleArrow, null))));
    })),
    React.createElement(StyleHiddenContainer, { ref: hiddenContainerRef }, items.map(function (item, index) { return (React.createElement(StyleItem, { key: index },
        React.createElement(StyleText, null, item),
        React.createElement(StyleArrow, null))); }))));
}`

describe('jsx-runtime', () => {
  test('detect', () => {
    const ast = babelParse(code, { sourceType: "module" });
    expect(getReactTag(ast.program)).toEqual('React')
  })
  test('wrap', () => {
    expect(reactInjectorLoader.bind({
      resourcePath: '/path/to/component.tsx',
      getOptions: () => ({
        encrypt: (key: string) => key
      })
    })(code)).toMatchSnapshot();
  })
})