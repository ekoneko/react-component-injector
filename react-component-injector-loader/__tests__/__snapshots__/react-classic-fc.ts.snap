// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx-runtime wrap 1`] = `
"import { reactComponentInjector } from \\"react-component-injector\\";
import React from 'react';

var Breadcrumb = function (_a) {
  return React.createElement(StyleWrapper, {
    className: className
  }, React.createElement(StyleContainer, null, items.map(function (item, index) {
    if (index !== 0 && index < hiddenCount + 1) {
      return null;
    }

    return React.createElement(StyleItem, {
      key: index
    }, React.createElement(StyleText, {
      onClick: createHandleClick(index),
      width: index < collapsedCount ? collapsedWidthPx + \\"px\\" : undefined
    }, item), React.createElement(StyleArrowWrapper, null, arrowIcon !== null && arrowIcon !== void 0 ? arrowIcon : React.createElement(StyleArrow, null)));
  })), React.createElement(StyleHiddenContainer, {
    ref: hiddenContainerRef
  }, items.map(function (item, index) {
    return React.createElement(StyleItem, {
      key: index
    }, React.createElement(StyleText, null, item), React.createElement(StyleArrow, null));
  })));
};

var $$WrappedBreadcrumb = reactComponentInjector.proxy(\\"/path/to/component.tsx#Breadcrumb\\", Breadcrumb);
export { $$WrappedBreadcrumb as Breadcrumb };"
`;
